### ISSUE TITLE
Citizens virtualization for large cities

### ISSUE THREAD

**Issue by dymanoid** [dymanoid/RealTime#165]
In the vanilla game, the citizens get spawned randomly, regardless of time of the day. Furthermore, there are only a few movement reasons that cause a certain spawn: like e.g. illness (find hospital) or need for goods (go shopping). Entertainment and _going to work_ are reasons that are throttled by the game depending on current number of spawned citizens and vehicles.

Since **Real Time** changes the _going to work_ behavior, this movement reason has now a very high priority. The only throttled reason left is entertainment. However, the throttling only occurs while a citizen is at home. This is because whenever a citizen is not at home, there are time restrictions to be applied (e.g. everyone must leave parks until night time, the workers must leave workplaces according to their work shifts etc). The only unrestricted place to be is "at home".

For large cities (>65k), this introduces a new kind of problems. The spawned citizens throttling relies on the _current number_ of spawned citizens. So, e.g. at 3 p.m., when almost all first shift workers are still at work (already despawned, because inside of buildings), the game allows to spawn lots of unemployed citizens (going e.g. to entertainment). At 6 p.m., when the first work shift ends, all the workers need to be spawned to get back home (or go shopping etc, according to their schedules). Since there is no throttling for citizens being at work, the number of spawned citizens quickly jumps to the maximum 65k. For very large cities, this even causes problems with citizens trying to leave their workplaces, because they cannot be spawned. So they have to wait at work for a "free slot".

This problem needs to be addressed somehow, but currently I have no ideas in which way.

**Comment by DaEgi01** [comment]
i guess this is where teleporting comes in in the original game behavior.
the only options i see is
a) increase the limit - have not checked if its technically possible or just too much work (best solution ^^ though performance will drop even further)
b) distribute the leaving of working places to a bigger time range - no solution since it only delays the problem (sucks but works for certain city sizes 65k < x < ?)
c) let them wait until a "citizen vessel" is available -> as is the case already (kinda sucks since it is basically a forced distribution with an open end, who knows, so cims will maybe never leave work ^^)
d) let them teleport (as vanilla i guess, kinda sucks, but it is what it is)
e) use another datastructure as an extension where the invisible cims are put into and where their position etc. is also calculated although they are not shown on the map. develop a mechanism to move cims from and to the extension depending on likelyhood of visibility to the player. -> (okeyisch, have not checked but unlimited trees mod does something like that i guess, only that trees are much easier than cims since their position is fixed and does not need to be calculated on update)

**Comment by Sipke82** [comment]
Perhaps reduce the allowance of unemployed workers to be spawned, so it's actually less busy during certain times of the day and especially at night....
In the middle of the night (3-4 AM), my 120k city still has around 55k cims spawned..

**Comment by dymanoid** [comment]
@DaEgi01, I see no difference between _d)_ and _e)_. Since the citizens are invisible, they don't walk, don't use public transport, don't drive. So there's no point in simulating their movement at all - just plain CPU load without any benefits.

@Sipke82, this doesn't help unfortunately. If there are 100k first shift workers that want to get home at 6 p.m., even completely disabed spawning of unemployed ones doesn't solve the problem.

The high number of citizen instances you observe at night can be explained: the dummy traffic, the shopping at night citizens, and the extended first shift workers are active at that time.

**Comment by dymanoid** [comment]
Possible solution - _smart teleporting_:

- calculate the teleporting probability according to current city population
- the starting point is always "at home"
- for every activity (work, shopping, entertainment), citizens either get spawned and move in their normal way, or teleport directly to the target
- a _teleporting_ flag is applied to the citizens that were teleported
- when moving further according to their schedules, citizens act according to the _teleporting_ flag (spawned ones can spawn further, _teleporting_ ones can only be teleported)
- this continues until the citizen is back home; then, the _teleporting_ flag is reset and will be recalculated according to the current situation

**Advantages:**

- apparently, the problem gets solved
- no ditched cars because of partly moving as spawned and partly getting teleported
- no skew towards workers (distributing the teleporting equally among employed/unemployed)

**Disadvantages:**

- more info needs to be saved in the save game (but maybe I can squeeze it in the current data structure)
- less realism
- probably less visible citizens e.g. in parks or hanging around by the buildings

**Comment by DaEgi01** [comment]
nah .. e) is much more complex in my mind than that.
they could become visible if you change the camera position for example.
covering distance still takes time, they still produce traffic and congestion.
they are just phased out visually while still contributing to the simulation.
its like in MMOs where players get "phased out" to other servers if it is too crowded.
its not easy to implement properly though, i would not want to do it, but look for another solution that is for sure ^^

**Comment by dymanoid** [comment]
Okay, now I get it. This is impossible with the current game's implementation, since you'd need to run all the instances simulation using the whole citizens buffer. Movement only is not enough because of public transport, collisions, etc etc. Even if that'd be technically possible, you'd need a monster overclocked machine to be able to run a, say, 400k city - the simulation is single-threaded (okay, with exception of path finding).

**Comment by DaEgi01** [comment]
i wonder ... isn't smart teleporting what the game already does?

**Comment by dymanoid** [comment]
Another idea: track the number of citizens that are _not at home_. (This number is not equal to the number of spawned citizens.) When the number of not-at-home citizens reaches the 65k threshold, no further citizens can leave their homes (regardless of their schedules).

**Advantages:**

- No teleporting
- Technically simple

**Disadvantages:**

- Large cities will function (mostly) as if they had only 65k citizens (visits count, income from shopping etc)
- Additional data still needs to be saved (but just one number)

**Comment by benzoll37** [comment]
Mods: only Real Time. City: 150k cims. Simulation mode: as in vanilla.

I see 2 hours tips, (not counting "lunch time": it's very small).
At the time of going to sleep (11pm) mod causes  2nd rush hour equal to the output of the first work shift (6pm). I suppose they are cims on the streets, in parks and shopping that return home to sleep.
It also begins in higher numbers because it collects the remains of traffic jams and active cims of the 1st rush hour (6pm).
This causes a prolongation until 3am or 4am of traffic jams and "life" in the city not realistic to the naked eye. _Maybe it would be good to eliminate this 2nd rush hour._ Then at 00am or 1am the rush hour of 6pm would dissipate and you would see a quiet night in the city.
I think with one rush hour, "Real Time" creates a realism that does not have vanilla game.

Note.- After upgrading to v 1.15 this has improved a bit: jams and "life" are reduced by one hour  before. .

**Comment by dymanoid** [comment]
@benzoll37, both unemployment and jobs assignment have nothing to do with this issue. It's only about spawned citizens.

**Comment by Sipke82** [comment]
i like option a:) from egi the most.... but i guess it will be impossible to increase the cim/vehicle limit. 

So basically, the game does not get more challenging past 65k cities, from a simulation game perspective.. in fact i think it will become less challenging, because the traffic (cims+vehicles) is will be more spread out over a larger area, and thus less congestion....

**Comment by DaEgi01** [comment]
that is exactly the case @Sipke82 
i call it the empty streets syndrome ;D
its funny how ppl showed cities >300k and were proud that they had no traffic issues ;(

**Comment by liiir1985** [comment]
I just took a look at the current implementation of CitizenManager. It is possible to tweek the limit of 65536 maximal citizen instances, but quiet hacky.( It used a type named Array16 to store all possible instances, which could only take 16bit indices) 
Is there any other problem of setting the limit to a higher value, despite the CPU and GPU cost?
Or, is it a solution for this issue if we could make it possible for the players to set this limit value through setting menu?

**Comment by liiir1985** [comment]
@Sipke82 Technically, it's the same as Citizen instance limit, and can be modified. Just... loads of methods needs to be patched

**Comment by dymanoid** [comment]
@liiir1985, I don't think it's even possible to increase the instances limits. Currently, there are many methods that do use hard-coded upper bounds while processing the arrays. How do you want to solve this? Please don't suggest the Harmony's "transpiler magic". Messing around with the IL code during run-time is the worst thing one could imagine. I know, some modders do use thins technique. But they maybe don't realize the danger of that approach.

Citizens limit is 1m by the way.

One more word about the multi-threading: the game doesn't support this, sorry. The current implementation has no built-in parallelism support, with a single exception for the path-finding thread.

**Comment by liiir1985** [comment]
@dymanoid It's actually a constant value which got compiled into IL code. yes, the only way to do it is to messing around the il code, but in a more clever manner. 
We could use code analysis library like Mono.Cecil(I had a lot of experience with such stuff, also at work. see my github repository- ILRuntime, which is a IL interpreter execution engine) to generate the patch automatically, instead of making it mannually, this would make it more reliable and consistent during future game updates.

The game doesn't support multi-threading, yes. But the calculation of the AI and all the simulation stuff(Logic is seprated from rendering) is already on another thread, rather than main thread. And synchronization lock is also exists(See SimulationManager, which will start a new Thread for all the simulation stuff), So basically we can modify the SimulationStep function of SimulationManager, to make the for loop multithreaded, or at least, to make the citzen/viehcle simulation multi threaded.

**Comment by dymanoid** [comment]
@liiir1985, that is exactly what I was talking about.

**Comment by liiir1985** [comment]
@dymanoid How about share riding? Will it solve the problem in some degrees? If the Citizen cannot be spawned, then call another citizen which has a viehcle to pick him up, then they move together. Maybe it's somehow more realistic than just teleporting the citizen directly

**Comment by liiir1985** [comment]
Let them ride virtually will still produce traffic to the area where this citizen lives, so maybe it will make the streets look more busy?

**Comment by dymanoid** [comment]
Traffic flow is calculated "by-car", not "by-citizen".

**Comment by liiir1985** [comment]
I know it's calculated by car. What I mean is, maybe it won't any car driving by otherwise, and now some one needs to pick him up, so it looks like a "spawned" car driving by.

**Comment by dymanoid** [comment]
I don't think it's a good solution in general. It won't solve the issue actually. For very large cities, every passenger car would need to pick up 3 passengers each time. Just imagine the traffic nightmare and the simulation performance drop in that case.

The feature itself is actually not bad. It adds some realism. But I see it rather in the "mass transit" domain - how does a particular citizen want to travel: by car, on foot, by public transport, or share a ride. So this is out of scope for **Real Time**.

**Comment by wbattel4607** [comment]
@dymanoid I like your smart teleporting idea. It's unfortunate that C:S has these limitations but I think that's probably the best way to go about it- at least for the time being. Maybe someday CO will improve on this given the improvements of the hardware of the average user since the game's launch- or if not, then perhaps in their successor to C:S they've mentioned before that'd eventually solve many of the technical limitations.

**Comment by aubergine10** [comment]
Maybe as population grows, increase +/- time randomisation factor of schedules, applied to all cims.

**Comment by dymanoid** [comment]
"Time zone" approach won't do, because it is not deterministic. We need a deterministic way to limit the number of "instantiated" citizens at a time.

@TMaekler, it's not about threading, the limitations are hard-coded in the game and cannot be overridden by user mods.

**Comment by cptalpdeniz** [comment]
One of the problem that needs to be addressed is which type of "cims" you want to get prioritized for instance slot. Basically, If you want to prioritize the tourist or other cims going outside for the day, you have to prioritize them for the slot and when its time to leave the work, workers have to wait. Or if you want to prioritize the workers, you can limit the instance slots during daytime or closer to workers leaving work to make room for them. However, this is just KISS approach to the problem. Which is not wanted.

Since we cannot alter how the game works and its limits, we can only maximize how we are working those limits. For instance, in 100K workers example given by @dymanoid, even if you have all 65K instance slots available, that is still not enough slots and you would end up with 35K workers whom has to wait till there's slot available.

If you want to implement something far deeper, which is what is wanted, I'd like to add few things to suggested ideas. They can be drastic changes, but I'm just trying to provide an idea to discuss it further and/or open up for more discussion.

**Change work/worker behavior**
As of now, with how the mod works is to make workers go to work and leave the work at a specific time (either default value or user selected). Even though this steps up the game in terms of realism from vanilla version, it's unrealistic. Not all the jobs are, for example, from 8 AM to 5 PM. By changing work behavior, you eliminate a chunk of workers from that _rush hour_ time frames. This is kind of what the mod does, but in a different way. Few examples for this are;

- Add randomness to the shift behavior. I understand that shift behavior is to make the game more realistic, however, not every job has to start exactly at that time. This decreases amount of people going to (or at least trying to) work at the same time. For instance, instead of shift starting at "work start time", it starts and finishes 1 hour later.
- Making the leisure areas, jobs, have shifts, more on weekends, later hours and at night instead of daytime. For example, there's no need for most of the leisure buildings to be open at 8AM, or Monday morning. You can add probability selector, what percentage of them open at work start time (first shift), just like in shift behavior, to make it more realistic by adding some randomness.
- Improve building's that are open 24/7, shift logic. Decrease the first shift percentage 
- Change Garbage processing to after 8PM to early morning to match real time.
- Remove lunch break completely

**Smart Teleporting**
Smart teleporting can be implemented to the game to help with cims stuck where they are because there is no free slot left. However, it is important to find a good ratio between how many cims are being teleported and how many of them are actually moving. 

- Since there is a instance limit, this also creates a traffic limit. Technically, every city created in Cities Skylines has a traffic limit. If you hit instance limit or close to it during those _rush hour_ times, you are also at traffic limit, which means the traffic cannot get worse since a cim with a car can't get spawned due to instance limit. 
- Instead of adding teleport flag to the cim, teleportation can be implemented in a dynamic fashion. For instance, it can depend on time of the day, how many cims are in line to spawn, what is the distribution of instances (how many are going to work, how many are out of shopping, how many are kids going to/leaving school, etc.) are and how game is treating these instances. In this way, you eliminate the probability of less people in parks and keeping that "rush hour" feeling.


**Adding Weather/Environment to the Simulation**
This is kinda fishy but small things like these help with both realism and performance. This does not have as big impact as other suggestions. Depending on the theme, weather (if its raining, cloudy, foggy, etc.), the movement of cims will change. Basically this would decrease number of cims going outside or having movements other than mandatory ones. If the weather is not nice, 

- Cims will stay at home during weekend
- Cims won't go to lunch breaks 
- Only small amount of tourists will actually do touristic activities during daytime.

All of the suggestions above should also differ in execution depending on the population of the city. As city grows, RT should increasingly disregard shift ratios and some other config settings. As @aubergine10 said it before, the bigger the city gets, the more Real Time should take control and start taking these measures to make it so that the performance doesn't get affected as much as it is right now.

**Comment by aubergine10** [comment]
I'm not sure what that last comment is about, it's sort of just reiterating  what's already been implemented or discussed elsewhere. For example, there are already  options to add variance to shift times, variance to the start ('on time') and end ('overtime') of shifts; people head indoors when it rains, and so on.

**Comment by dymanoid** [comment]
I just wanted to write the same comment as  @aubergine10. Thanks @cptalpdeniz for your effort, but in fact you don't propose any solution. The whole probability-related stuff is non-deterministic and won't solve the problem. The teleporting cannot be dynamic because there will be ditched cars everywhere. And so on. I didn't find any suggestion in your post that would help solve the issue, unfortunately.

**Comment by dymanoid** [comment]
This would eliminate the commuter traffic completely and make the game totally boring.
But the problem is not necessarily related to those citizens. In large cities, there is always a probability to hit the citizen instances limit because of citizens' schedules that **Real Time** implements. Please read the first post. The vanilla game can effectively throttle the number of spawned citizens because there are only few reasons for a citizen to be _necessarily_ spawned. **Real Time**'s schedules change that so there are _many_ reasons for _necessary_ spawning.

**Comment by TMaekler** [comment]
I understand. My line of reasoning was to create a new type of citizen (which basically acts like the normal citizen), but can additionally be spawned alongside normal citizens (they should for example fill up public transport slots, drive around in cars etc.). Thereby circumventing the ingame limits of 65k spawned citizens.

**Comment by aubergine10** [comment]
Add a scaling factor to the cims and vehicles? Once limit is reached, they start getting physically bigger...

**Comment by dostillevi** [comment]
As I see it, there are two main problems caused by the citizen and vehicle limits:

1. Citizens get stuck and can't get to where they need to go while waiting for a slot to be open.
2. The games becomes unrealistic when most solutions to #1 are applied, namely teleporting.

The only solution I see to #1 involves teleporting. The cap on citizens has already been discussed and no solution is available. However for appearances sake, could groups of citizens be made that are treated by the game as one citizen but hold stats values for several, or even are simply visually more than one citizen? 

Alternately, not all citizens the player sees need to be following the simulation. As long as what they're doing seems realistic, they can just be fancy props. The key is to have enough citizens within the player view doing simulation things that those standing around, walking around the block forever, etc aren't so obvious. This might require writing some simulation code outside the main simulation or creating a new class of object, and I'm not sure how feasible that would be.

So that leaves #2, if citizens are teleporting, how can the traffic simulation be preserved? As I understand it, the traffic simulation is done on a "per vehicle" basis. What if vehicles could be grouped together with multiple vehicles temporarily being treated as one when they're traveling together in groups in the same lane? Two vehicle groups could act like an articulated bus as long as they're both headed in the same direction.

**Comment by timapelov** [comment]
Support of the above idea - I suggest several cims, moving from one common point to another common point, to unite in groups and simulate these groups. Then have us will 65K groups, in which can be 2-3-4-5 people. And for statistics (transport, entertainment, job) to consider them separately, it's easy, because the place and transport for the group is one. When the groups get to the point, you can combine them into a new group or calculate as separate cims. In real life-many colleagues go somewhere together after work, family members often move in groups (joint purchases in the store, joint vacation). Indeed, in C:S all the cims alone and the car alone for one cims. Groups can be visualized as multiple people walking side by side, there will be one model for one group and one travel simulation . What do you think @dymanoid ? 

Additional question. Do a lot of computer resources spent on the calculation of traffic without cim (trucks, transport, municipal services). If so, is it possible to disable the simulation of urban services (let's say garbage) manually? In order to have more computer resources for cims.

* sorry for the translator*

**Comment by dymanoid** [comment]
I will be joining a Modders Meetup at Colossal Order in June 2019. I'll discuss this topic with the game developers directly. Maybe they will have an idea about how to solve this.

**Comment by RenaKunisaki** [comment]
The solution is for them to change all those 16-bit indices to 32-bit. I hope you'll bring that up. Maybe it was a good way to reduce memory/CPU usage, but 4 years later our computers are a little more powerful.

As for temporary workarounds, it seems like one solution would be to impose limits per type of citizen. Eg a maximum of 20000 unemployed can be spawned at once, 20000 employed, 20000 tourists. Then, those limits can gradually change with the time of day; eg from 3am to 8am the limit for Employed gradually rises, and the limit for others gradually drops, until at 8am the Employed limit is maybe 50000. The key is the limits change gradually.

At all times the limits should add up to around 60000, which leaves you with a buffer of 5535, that allows citizens to get to places when they're over the limit. (eg if 20000 unemployed are currently spawned, and that limit starts to drop, then some unemployed who are near home or near their destination will hurry up - maybe even start physically moving faster - to try to bring the number back under limit quickly.)

The limits could also be adjusted by region/district; eg if a district has a lot of employed people, then it will generally have higher limits for number of employed people. (If a district is at its limit, people won't decide to go there, but people who are already en route might still go, prompting someone else to leave; or, they might switch to a different destination.)

Eventually, if you run up against the hard 65535 limit, you could fall back to despawning/teleporting both the citizen and their car, or forcing them to wait. It's unrealistic, but shouldn't be too noticeable. If you look closely you already see other unrealistic behaviour at that scale (cars driving through eachother, people pulling bicycles out of their pockets...); at some point you have to accept that a game won't be 100% realistic and you're going to see such quirks occasionally, especially when you're pushing the limits of its design.

Another thought: do people count toward the spawn limit while in vehicles? Maybe you can despawn people and just let empty vehicles drive around? Draw some models inside to look like people but don't actually have a person inside?

**Comment by aubergine10** [comment]
> Maybe it was a good way to reduce memory/CPU usage, but 4 years later our computers are a little more powerful.

There are still vast numbers of people on potato computers, particularly laptop users.

**Comment by willbattel** [comment]
> > Maybe it was a good way to reduce memory/CPU usage, but 4 years later our computers are a little more powerful.
> 
> There are still vast numbers of people on potato computers, particularly laptop users.

This is an argument as old as time itself. There will always be personal computers considered "potato", but the standards for which they qualify does increase over time. The average computer today is certainly more powerful than the average 4 years ago- however, as you said, there are still a significant enough number of users on insufficiently powerful devices that hold back more compute/memory-intensive game development.

Additionally, increasing memory address size isn't a trivial change. At that point, CO might be more inclined to just make C:S 2- but that isn't likely to happen until their DLC development starts seeing diminishing returns.

Some day.

**Comment by RenaKunisaki** [comment]
I got the impression from previous comments that it would be a pretty simple change, but one that has to be done at source-code level, not something that a mod can do. Is that not the case?

**Comment by willbattel** [comment]
> I got the impression from previous comments that it would be a pretty simple change, but one that has to be done at source-code level, not something that a mod can do. Is that not the case?

Maybe; It's impossible to say without seeing their code base. _Usually_ it would be a major overhaul but that's not always the case. They may have developed it such that it'd be an easy change.

**Comment by aubergine10** [comment]
Even if it is a trivial change, like updating a single constant and then recompiling, it doesn't necessarily mean its a feasible change - there will be so much stuff linked to that value that, should it be increased, you're effectively multiplying the computations taking place across large swathes of the codebase.

Case in point, in a recent pull request review on the TMPE mod, this cropped up:

> Assuming every source lane has `2` target lanes on average: `n * 2` extra computations needed (max. `262144 * 2 = 524288`) which gives us a total maximum of `2*n*n + n*2 = 137,439,477,760` required calculations.

That's the outcome of an increase from `1` to `2`. Simple change, big ramifications.

**Comment by aubergine10** [comment]
Doesn't solve the core issue, but Quboid's [Cargo Hold Fix](https://steamcommunity.com/sharedfiles/filedetails/?id=1721492498) mod has option to turn off dummy traffic, which might free up some vehicle slots on mid-sized cities.

**Comment by supermerill** [comment]
Do you know the number of car & pedestrian instance in an average (good) city?
Maybe teleporting pedestrians to their next hop (with a timer, a bit like when they are at work, but instead they are "waiting at the station") can free up some agent slots. But it may be a big change because now public transports have to check the waiting agents & the new "ghost queue".

Anyway, it's pretty useless if pedestrians count < car count.

**Comment by RenaKunisaki** [comment]
Doesn't the game already do that?

**Comment by dymanoid** [comment]
So here are the results of the discussions with Colossal Order.

Changing the internal data structures is unfortunately not possible anymore. This is the very game's core, so it would require a lot of changes in the code and retesting all the functionality. Frankly, such a change is just too overwhelming, so Colossal Order will not touch that part.

We also found no other possibilities to circumvent the issue. Colossal Order planned their game in such way that the simulation workload is evenly distributed across the whole game time. With **Real Time**, we are kind of going against the original game architecture.

So we are forced to implement workarounds. I think, the "smart teleporting" approach is probably the best way we can go in this case.

**Comment by RenaKunisaki** [comment]
That's unfortunate.

One thing I've thought of, but I don't remember if it was mentioned: batching citizen instances, and "fake" citizens.

Example: two people are walking along the same path and won't be branching off for a while. You replace them both with a single person, whose model is actually both of their models, so it still appears as two people. When clicking on that instance, or when they reach the point where they're going to diverge, then you respawn the actual people. Since it's mostly a visual trick, it could even be done depending on camera distance.

Another case is places like bus stops, where people stand around a lot. You could despawn them when they're just standing there, replacing them with an object that looks like them, and displays their info when clicked, but isn't actually a citizen instance. You still keep track of the actual person in a separate buffer, so that you can respawn/find them as needed (eg if player clicks on their vehicle and looks up their position).

**Comment by DaEgi01** [comment]
the idea with stations is pretty nice :)

**Comment by tomvaneyck** [comment]
After reading the suggestions in this thread, I feel like the best solution to this problem may be a combination of _smart teleporting_ and a distribution over the cims activities.

**Smart teleporting**
I would make a change to the _smart teleporting_ idea: instead of using a flag (which is not taking into account the amount of spawned cims when returning home) decide when leaving a place (home, work, shop ...) if the "busyness" allows to spawn, otherwise teleport.

One way to decide this is to use a chance of spawning, inversely proportional to the free spots left. This behavior is deterministic, because no cims will spawn when the spawn limit is reached as the the chance would equal zero. This behaviour is also fair, in the sense that every cim has an equal chance of being spawned.

**Distribution accross work and other activities**
Like mentioned before, a higher chance of spawning commuting cims during rush hours delivers a more realistic expirience. However, in big cities, depending on the updating order of the simulation, cims with other destinations than work may take up precious spawn places. This may lead to an unrealistic distribution of traveling cims.

Having a distribution enforced by chances when spawning will make sure that during rush hour, for instance, 80% of traveling sims will be commuters. This distribution has to change based on time of day. This behaviour can also be considered deterministic for larger numbers of citizens, so this behaviour can be delayed untill the city has reached the amount of citizens equal to the limit of spawned cims.

**Conclusion**
Because of the reactive nature of the smart teleporting to the current amount of spawned cims, the limit will not be exceeded. The chances used in this method will behave deterministically and will lead to a realistic distribution when use for high enough numbers of citizens. Still, a thousand citizens would already be enough for this to be true.
This method will also not take a lot of extra computational time, because only simple chances are calculated, so I do not expect a real performance impact.

I really think this would work, but it is late and I have only quickly looked over the codebase. If anything is unclear, please say so and I will try to clarify.

**Comment by dymanoid** [comment]
If we'd decide about teleporting "on both sides" (there and back), we could get undesirable side effects like e.g. ditched cars or improper public transport usage.

Imagine a cim gets spawned and drives to the work building. On return, there is no slot, and the cim gets teleported home. Now, the car is ditched at the work building - the cim won't find the car anymore.

Or, imagine the cim gets teleported to the work building. On return, there *is* a slot, and the cim gets spawned. Since there's no car, the cim will use public transport (or move on foot), preventing other cims to spawn which used their cars to get to work and thus causing those cims to be teleported home (producing more ditched cars).

Not that easy :)

**Comment by azabost** [comment]
I have no experience developing a C:S mod so you can treat my comment as a joke, but it bothers me so much I just have to ask 😄 Did you consider if this issue could be fixed by implementing a "carpooling" somehow? I.e. two cims living close to each other can travel together in a single car (or, technically, there can be a single cim-car travelling and the other teleporting with it). Or they could share a car so that the first cim drops off at his home and the second cim continues to his home. Sorry if I'm talking gibberish 😄

**Comment by matfax** [comment]
The real question is if C:SL2 will solve the performance issues and enroll a sophisticated and multithreaded crowd AI. It's definitely possible but will they invest in this topic or focus on other features? Updating to a newer Unity version won't fix the model. Transport Fever 2 is a new alternative to C:SL2 with regards to crowds and vehicle simulation and yet, they have the same issues as C:SL2 does and as Cities XL had. Don't go big, your fps drops otherwise. So I wouldn't be too optimistic about it. And it's completely unnecessary to recalculate everything each frame if you think about it. There are so many ways to circumvent it. Using adaptive schedulers being the most common one, I think. This can be further extended to a prioritization of different scheduled tasks in a functional way. There is no need to precisely calculate AI objects separately if you never look at them. Only calculate what's inside the view and approximate the remaining stuff over a whole second instead of each frame. Pathfinding can be approximated and generalized as well. This topic has been discussed for TMPE [here](https://github.com/CitiesSkylinesMods/TMPE/issues/316). There's the same potential for public transport (not for personal vehicles) or even for crowds. If a certain group of objects is expected to behave similarly, you can save time by generalizing them.

**Comment by Sipke82** [comment]
One small idea I had with regard to teleporting and cars:

Today with the More Vehicles mod (amazing O-o ) we have the following limits:
64k vehicles
64k cims instances
I'm my experience the cim instances limit get reached first.

So perhaps we can let the "car drivers" be teleported into their cars and from their cars to a in a building.

**Comment by pleaserespond** [comment]
I'm just a player mostly but would like to give my 2c. Obviously we have to accept the hardcoded 64k limit for simultaneous spawned agents, so the engine fundamentally cannot provide a fine-grained simulation of a large city. Having said that all we can do is provide compromises that keep the spirit of Real Time mod within the limits of the engine. Here are a few ideas I've had, I read the entire thread a few times so I'm fairly certain these weren't suggested, apologies if I'm repeating old suggestions.

1. Randomise the time cims leave. Yes, the cim has to be at work at 9, but really we all know that +/-15m is acceptable. By applying a jitter function to the spawn time we'll hit the spawn limit with different cim distributions each day, providing an overall traffic distribution that approximates the full load over several days. This is similar to the time zone idea, but without being hardcoded so your road network has to take every possible distribution.
2. Impromptu carpooling. If a cim wants to go from home to somewhere else, but can't spawn, find a different cim that is making a similar journey and teleport the bodyless cim to their destination when the carpool cim reaches theirs. For the return trip do the same. It's still possible there will be no carpool to take, but it's something.
3. Do not fully simulate cims not in cars. A cim taking public transport won't affect the road network, since the actual vehicle is already spawned. You could calculate which routes the cim would take and teleport them to their destination once the final leg of the journey is complete. If the cim decides to drive part of the way, make them wait either for a slot or for a carpool. This in effect is a shadow simulation for cims that walk or take public transport, since they don't participate as much in the road network.
4. If a cim waits too long for any option to be available (let's say 1 in-game hour), calculate roughly the expected time their trip would take and just teleport them. Not ideal, but the city must keep working.

**Comment by dymanoid** [comment]
Thanks for your suggestions! I can comment on some of them.

1.  This is already happening. Real Time does randomize the time for the citizen schedules. Additionally, the gam randomizes them even more. If Real Time says "this citizen would leave their home at 9:00 AM", the game might simulate that citizen e.g. only at 9:30 AM, so the citizen leaves later than planned. This is because the game simulates the citizens in chunks, so all them are distributed across the simulation cycle which is (depending on Real Time speed setting) something between 1 min. and 40 min. of the in-game time.
2. This was discussed above. It doesn't really help: the routes change (roads get changed by the player, vehicles can be despawned etc.), and searching for a similar route would mean a *significant* performance impact, especially in large cities. Knowing only the destination is not enough, because citizens choose different ways of travelling (on foot, public transport, bicycle, car).
3. I don't quite understand this point. Currently, citizens moving by public transport will not be spawned. They only get spawned when they exit the public transport vehicle, and we need to spawn them.
4. 1 hour is definitely too few, but this is actually the general idea of the teleporting. It will always be the last fallback solution (and maybe the only feasible solution...)

**Comment by pleaserespond** [comment]
My thinking was mostly about how to best choose who gets teleported when not enough slots are available. It's a kind of scheduling problem, except we have 65536 "cores" and processes are not preemptible and magically autocomplete if they don't run.

For 2, I don't think it will be that hard - you can keep track of every (start, end) points for the route of every vehicle on the road in a quadtree structure, so you can quickly answer queries like "is a vehicle making the journey from a point close to A to a point close to B right now". If yes, you can pretend the cim who wants to go from A to B actually carpooled with the existing vehicle and teleport them. Yes, this cim could have chosen a different route and yes, a simple geometric distance doesn't account for the fact that the cim might have to walk more than they're willing, but teleporting them now allows a different cim to spawn whose journey is entirely unaccounted for.

You don't even need to record which actual vehicle is making the trip, just a pair of (start, end) points and (start, end) times. You add each journey as it is completed - because we'll teleport the cim, we need the journey to be already complete - and delete ones that ended too long ago.

I think this option will be most useful for cims going to/from work or school - a lot of times several cims from one neighbourhood want to go to roughly the same workplace, so this allows them to be grouped in one vehicle.

For 3, I meant to prefer teleporting cims if their journey is entirely with public transport and walking, subject to how oversubscribed the system is. Or maybe I should have formulated it as "deprioritise spawning cims who will not drive when a free slot is available". 

I see these as cases where teleporting the cim doesn't change the simulation too much, so the waiting time for teleporting the cim can be shorter than usual, which means few less cims waiting.

**Comment by kittenchilly** [comment]
Would it be possible to have a teleporting system that calculates the average time it would take for someone to make it to a certain destination and just teleport them after that time is up? Or would that be too expensive?

**Comment by dostillevi** [comment]
After some long thought, I had an idea, but it depends on the data structures used by Cities: Skylines to store citizen data. My idea is to replace what I assume is a single addressable array of 65536 citizens with a set of 10 addressable arrays, 0-9. In order to store or retrieve citizen data, the first digit of the citizen ID would reference the array, while the remaining digits would specify the location within the array. 

In order for this to be useful, the limitation of 65536 must exist only in the storage array, not in any formulas that query the array. Those formulas would look up with IDs that look like 100123 (for the 124th entry in the second array). Presumably there isn't a hard coded limit to prevent these variables from accepting a 6 digit numeric value.

**Comment by Evengard** [comment]
The limitation is actually in the ID - it can't tace a number greater than 65535, aka it's an UInt16.

**Comment by sairam4123** [comment]
So, what's the plan, how will this get implemented? I'm working on a similar game and I'd like to learn more on this. 
Thanks!

**Comment by isaac-prog** [comment]
For larger cities, a huge handful of Cims are seen waiting around Outside connections (plane,bus,train,ferries) and they take up alot of those citizen agent instances. If there is a way to make them mass spawn roughly the same time as their intercity transport, it would definitely  save lots of those citizen agent instance and optimised intercity transport to the fullest.

**Comment by dymanoid** [comment]
> Thus i wish to ask if 4 cims enter a vehicle, does the agent count reduce from 4 to 1, or increases to 5 (4cims+1car)?
When 4 cims enter a vehicle, the number of cim instances decreases by 4 (because all 4 instances despawn). The number of vehicles increases by 1 if the car is pocket-spawned at that moment; otherwise, the car instance status is changed from "parked" to "moving" and thus the number of vehicles remains the same.

**Comment by isaac-prog** [comment]
so...is there a solution for the outside connections mentioned above? :)

**Comment by algernon-A** [comment]
Work is currently underway on a project that will extend the available citizen instances beyond the 65K limit; when that happens (props already done, buildings next, citizens after that) I'll make sure that this mod is updated to use the extended manager, and then this issue will be resolved.

**Comment by algernon-A** [comment]
> I would like to clarify if active vehicles is a subset of citizen instance or stored seperately. ie to say if have 16k active moving vehicles, does that mean it consumes 16k citizen instance or 0 citizen instance?

Missed this response - vehicle limits are separate to Citizen Instances.  The More Vehicles mod already unlocks that particular limit by expanding it to the full 65K 16-bit limit.

**Comment by ruzbeh0** [comment]
Here is a possible workaround to this problem.

Instead of increasing the citizen istance limit, a solution would be to assign a weight to each citizen.  The main idea is that the citizens are not a represention of the whole population but a sample of it. For example, each citizen instances could represent 2 citizens.

I haven't actually used the Real Time mod yet, but I had a city that was getting too big and I was able to implement this idea by using several other mods. 

I used realistic population to reduce the number of households and jobs in each building by half, used another mod to reduced the capacity of transit vehicles by half, reduced the capacity of outside connections by half, and doubled my income to compensate for the lost in revenue. The game is showing that I now have about half of the popution that I did before, but in practice it represents the double of what is shown - but without passing the limit of citizen instances. The only thing I wasn't able to adjust is the number of vehicles on the roads, which need to be doubled. But since the vehicle limit is much higher than the citizen instance that shouldn't be a problem.

**Comment by isaac-prog** [comment]
i was wondering if it is possible to replace some 1 cim character with 2 cim characters, ie 1 red shirt guy with 1 red shirt guy and 1 blue shirt guy. so these 2 guys will go to the same destination, take the same path and so on, but the game treats them as only 1 cim character. i know this is more about creating a "2 cims citizen" asset in the citizen workshop, but wondering if this could at least make the city look more lively with "more" cims.

**Comment by dymanoid** [comment]
@isaac-prog  Real Time doesn't restrict the number of spawned citizens. The problem is the opposite. In big cities and at certain times, Real Time needs to spawn more citizens than the game is able to do. So spawning 2 guys instead of one, as you suggest, will make the situation even worse than it is now.

**Comment by eberkain** [comment]
I admit i have not read this entire thing so this may have already been suggested.  Can you make a virtual rideshare, so if there are 4 people that all want to leave work at the same time, you spawn one out and then save the other 3 in the background. The spawned cim then drives to the homes of the other cims in order and deposits them at their respective locations before driving him self home. In theory it could scale to any size if you dont restrict the people that can pile in one car, at the cost of making commutes longer. 

You would have to despawn vehicles for anyone doing the rideshare just to be safe, but if you are hitting the citizen cap then you are probally hitting the vehicle cap anyhow.
